plasmapy/diagnostics/proton_radiography.py

```diff
@@ -0,0 +1,955 @@
+        be entering the grid (to save time)
+        """
+        # Distance from the source to the nearest gridpoint
+        dist = np.min(np.linalg.norm(self.grid - self.source, axis=3))
+
+        # Time for fastest possible particle to reach the grid.
+        t = (dist / self.v0).to(u.s)
+
+        self.r = self.r + self.v * t
+
+    def _advance_to_detector(self):
+        r"""
+        Once all particles have cleared the grid, advance them to the detector
+        plane.
+
+        This step applies to all particles, not just those that will hit the
+        grid.
+        """
+        dist_remaining = np.dot(self.r, self.det_n) + np.linalg.norm(self.detector)
+
+        v_towards_det = np.dot(self.v, -self.det_n)
+
+        # Time remaining for each particle to reach detector plane
+        t = dist_remaining / v_towards_det
+
+        # If particles have not yet reached the detector plane and are moving
+        # away from it, they will never reach the detector.
+        condition = np.logical_and(v_towards_det < 0, dist_remaining > 0)
+        ind = np.nonzero(np.where(condition, 0, 1))[0]
+        self.r = self.r[ind, :]
+        self.v = self.v[ind, :]
+        self.nparticles_grid = self.r.shape[0]
+        t = t[ind]
+
+        self.r += self.v * np.outer(t, np.ones(3))
+
+        # Check that all points are now in the detector plane
+        # (Eq. of a plane is nhat*x + d = 0)
+        plane_eq = np.dot(self.r, self.det_n) + np.linalg.norm(self.detector)
+        assert np.allclose(plane_eq, np.zeros(self.nparticles_grid), atol=1e-6)
+
+    def _place_particles(self):
+        r"""
+        For each particle, find the indicies of the nearest field assuming that
+        the fields are placed on a regular grid
+        """
+        # Interpolate the grid indices that best match each particle position
+        i = self.interpolator(self.r[self.gi, :].si.value)
+
+        # Store the grid positions
+        self.xi = i[:, 0].astype(np.int32)
+        self.yi = i[:, 1].astype(np.int32)
+        self.zi = i[:, 2].astype(np.int32)
+
+    def _adaptive_dt(self, B):
+        r"""
+        Calculate the appropraite dt based on a number of considerations
+        including the local grid resolution (ds) and the gyroperiod of the
+        particles in the current fields.
+        """
+        # If dt was explicitly set, ignore this fcn
+        if self.dt is not None:
+            return self.dt
+
+        # Compute the timestep indicated by the grid resolution
+        # min is taken for irregular grids, in which different particles
+        # have different local grid resolutions
+        gridstep = (np.min(self.ds) / self.v0).to(u.s)
+
+        # If not, compute a number of possible timesteps
+        # Compute the cyclotron gyroperiod
+        Bmag = np.linalg.norm(B, axis=1)  # B is [nparticles,3] here
+        gyroperiod = (2 * np.pi * const.m_p.si / (const.e.si * np.max(Bmag))).to(u.s)
```

Comment:
I'm wondering if we'd want to make this a `plasmapy.formulary` function since I expect it'll come up fairly often.

Reply:
It would probably be easier to do something like...

```python
gyroperiod = (1. / gyrofrequency(np.max(Bmag), particle="p", to_hz=True)).decompose()
```

Reply:
It comes up often enough that I wouldn't mind a thin wrapper for it, just for the convenience, even with a oneliner docstring: inverse gyro frequency, see here: gyro frequency.

Reply:
That sounds like a good solution to me: should we open a separate PR for that or can I just add that wrapper function as part of this PR?